### PR TITLE
[selenium-webdriver] Add virtual authenticator library

### PR DESCRIPTION
**Pull request by TamsilAmani**
Please fill in this template.

Motivation : https://github.com/SeleniumHQ/selenium/issues/11184

- [x] Use a meaningful title for the pull request. Include the name of the package modified.
- [x] Test the change in your own code. (Compile and run.)
- [x] [Add or edit tests](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#my-package-teststs) to reflect the change.
- [x] Follow the advice from the [readme](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#make-a-pull-request).
- [x] Avoid [common mistakes](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#common-mistakes).
- [x] [Run `npm test <package to test>`](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#running-tests).

Select one of these and delete the others:

If adding a new definition:
- [x] The package does not already provide its own types, or cannot have its `.d.ts` files generated via `--declaration`
- [x] If this is for an npm package, match the name. If not, do not conflict with the name of an npm package.
- [ ] Create it with `dts-gen --dt`, not by basing it on an existing project.
- [ ] Represents shape of module/library [correctly](https://www.typescriptlang.org/docs/handbook/declaration-files/library-structures.html)
- [ ] `tslint.json` [should contain](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#linter-tslintjson) `{ "extends": "@definitelytyped/dtslint/dt.json" }`, and no additional rules.
- [ ] `tsconfig.json` [should have](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#tsconfigjson) `noImplicitAny`, `noImplicitThis`, `strictNullChecks`, and `strictFunctionTypes` set to `true`.

### CODE REVIEW

**Comment by DangerBotOSS** [comment]
Inspecting the JavaScript source for this package found some properties that are not in the .d.ts files.
The check for missing properties isn't always right, so take this list as advice, not a requirement.
## selenium-webdriver ([<kbd>unpkg</kbd>](https://unpkg.com/browse/selenium-webdriver@latest/))
was missing the following properties:
1. ThenableWebDriver
<p align="right">
  Generated by :no_entry_sign: <a href="https://danger.systems/js">dangerJS</a> against cb4d2386f254a6186a9095e864dbb888d9774227
</p>

File: types/selenium-webdriver/lib/virtual_authenticator.d.ts [hunk 1]
```diff
@@ -0,0 +1,108 @@
...
+        signCount: number
+    )
+
+    id(): Uint8Array;
+
+    isResidentCredential(): boolean;
+
+    rpId(): string;
+
+    userHandle(): Uint8Array;
```
Comment (SupernaviX) [inline types/selenium-webdriver/lib/virtual_authenticator.d.ts:78]:
```suggestion
    userHandle(): Uint8Array | null;
```
The docs say that `userHandle` [may not be defined](https://w3c.github.io/webauthn/#credential-parameters), and this implementation [does apparently use null sometimes](https://github.com/SeleniumHQ/selenium/blob/trunk/javascript/node/selenium-webdriver/lib/virtual_authenticator.js#L150-L155).

File: types/selenium-webdriver/lib/virtual_authenticator.d.ts [hunk 2]
```diff
@@ -0,0 +1,108 @@
...
+/**
+ * A credential stored in a virtual authenticator.
+ * @see https://w3c.github.io/webauthn/#credential-parameters
+ */
+export class Credential {
+    constructor(
+        credentialId: Uint8Array,
+        isResidentCredential: boolean,
+        rpId: string,
+        userHandle: Uint8Array,
```
Comment (SupernaviX) [inline types/selenium-webdriver/lib/virtual_authenticator.d.ts:67]:
```suggestion
        userHandle: Uint8Array | null,
```
The docs say that `userHandle` [may not be defined](https://w3c.github.io/webauthn/#credential-parameters), and this implementation [does apparently use null sometimes](https://github.com/SeleniumHQ/selenium/blob/trunk/javascript/node/selenium-webdriver/lib/virtual_authenticator.js#L150-L155).

File: types/selenium-webdriver/lib/virtual_authenticator.d.ts [hunk 3]
```diff
@@ -0,0 +1,108 @@
...
+    /**
+     * Creates a resident (i.e. stateless) credential.
+     * @param id Unique base64 encoded string.
+     * @param rpId Relying party identifier.
+     * @param userHandle userHandle associated to the credential. Must be Base64 encoded string.
+     * @param privateKey Base64 encoded PKCS
+     * @param signCount initial value for a signature counter.
+     * @returns A resident credential
+     */
+    static createResidentCredential(id: Uint8Array, rpId: string, userHandle: Uint8Array, privateKey: string, signCount: number): Credential;
```
Comment (SupernaviX) [inline types/selenium-webdriver/lib/virtual_authenticator.d.ts:93]:
```suggestion
    createResidentCredential(id: Uint8Array, rpId: string, userHandle: Uint8Array, privateKey: string, signCount: number): Credential;
```

This method isn't static, [the tests](https://github.com/SeleniumHQ/selenium/blob/000c2cf10d522a90bb765e6ff6882114eefcadc4/javascript/node/selenium-webdriver/test/lib/credentials_test.js#L145) assert that.

File: types/selenium-webdriver/lib/virtual_authenticator.d.ts [hunk 4]
```diff
@@ -0,0 +1,108 @@
...
+
+    /**
+     * Creates a non-resident (i.e. stateless) credential.
+     * @param id Unique base64 encoded string.
+     * @param rpId Relying party identifier.
+     * @param privateKey Base64 encoded PKCS
+     * @param signCount initial value for a signature counter.
+     * @returns A non-resident credential
+     */
+    static createNonResidentCredential(id: Uint8Array, rpId: string, privateKey: string, signCount: number): Credential;
```
Comment (SupernaviX) [inline types/selenium-webdriver/lib/virtual_authenticator.d.ts:103]:
```suggestion
    createNonResidentCredential(id: Uint8Array, rpId: string, privateKey: string, signCount: number): Credential;
```
This method isn't static, [the tests](https://github.com/SeleniumHQ/selenium/blob/000c2cf10d522a90bb765e6ff6882114eefcadc4/javascript/node/selenium-webdriver/test/lib/credentials_test.js#L108) assert that.

File: types/selenium-webdriver/lib/virtual_authenticator.d.ts [hunk 5]
```diff
@@ -0,0 +1,108 @@
...
+ */
+export interface ITransport {
+    BLE: string;
+    USB: string;
+    NFC: string;
+    INTERNAL: string;
+}
+
+export const Protocol: IProtocol;
+export const Transport: ITransport;
```
Comment (SupernaviX) [inline types/selenium-webdriver/lib/virtual_authenticator.d.ts:22]:
```suggestion
/**
 * Protocol for virtual authenticators
 * @enum {string}
 */
export enum Protocol {
    CTAP2 = 'ctap2',
    U2F = 'ctap1/u2f'
}

/**
 * AuthenticatorTransport values
 * @enum {string}
 */
export enum Transport {
    BLE = 'ble',
    USB = 'usb',
    NFC = 'nfc',
    INTERNAL = 'internal',
}
```

These look like they're meant to be used as enums

File: types/selenium-webdriver/lib/virtual_authenticator.d.ts [hunk 6]
```diff
@@ -0,0 +1,108 @@
...
+export class VirtualAuthenticatorOptions {
+    constructor();
+
+    getProtocol(): string;
+
+    setProtocol(protocol: string): void;
+
+    getTransport(): string;
+
+    setTransport(transport: string): void;
```
Comment (SupernaviX) [inline types/selenium-webdriver/lib/virtual_authenticator.d.ts:37]:
```suggestion
    getProtocol(): Protocol;

    setProtocol(protocol: Protocol): void;

    getTransport(): Transport;

    setTransport(transport: Transport): void;
```

Making those two consts into enums means they can be used as types here

**Review by SupernaviX — COMMENTED** [review]
I think there are a few issues with these types. Mostly some methods are incorrectly static, but also some enum-like values could just be enums.

File: types/selenium-webdriver/lib/virtual_authenticator.d.ts (hunk 3 above)
Comment (TamsilAmani) [reply types/selenium-webdriver/lib/virtual_authenticator.d.ts:93]: This method has been made static now with [MR](https://github.com/SeleniumHQ/selenium/pull/11182/files). The [tests](https://github.com/SeleniumHQ/selenium/pull/11182/files) have also been modified.

File: types/selenium-webdriver/lib/virtual_authenticator.d.ts (hunk 4 above)
Comment (TamsilAmani) [reply types/selenium-webdriver/lib/virtual_authenticator.d.ts:103]: This method has been made static now with [MR](https://github.com/SeleniumHQ/selenium/pull/11182/files). The [tests](https://github.com/TamsilAmani/selenium/blob/d7453d1bd0a5cbdc7fea346433dc1ecd589e974c/javascript/node/selenium-webdriver/test/lib/credentials_test.js#L106) have also been modified.

File: types/selenium-webdriver/lib/virtual_authenticator.d.ts (hunk 1 above)
Comment (TamsilAmani) [reply types/selenium-webdriver/lib/virtual_authenticator.d.ts]: Change incorporated.

File: types/selenium-webdriver/lib/virtual_authenticator.d.ts (hunk 6 above)
Comment (TamsilAmani) [reply types/selenium-webdriver/lib/virtual_authenticator.d.ts]: Change incorporated. Thank you @SupernaviX

File: types/selenium-webdriver/lib/virtual_authenticator.d.ts (hunk 2 above)
Comment (TamsilAmani) [reply types/selenium-webdriver/lib/virtual_authenticator.d.ts]: Change incorporated.

File: types/selenium-webdriver/lib/virtual_authenticator.d.ts (hunk 5 above)
Comment (TamsilAmani) [reply types/selenium-webdriver/lib/virtual_authenticator.d.ts]: Change incorporated.

**Review by SupernaviX — APPROVED** [review]
Looks good!

**Review by sandersn — APPROVED** [review]

**Comment by TamsilAmani** [comment]
Ready to merge

**Pull request merged**